dvc/stage.py

```diff
@@ -789,8 +789,34 @@ def _check_duplicated_arguments(self):
 
     def _run(self):
         self._check_missing_deps()
-        executable = os.getenv("SHELL") if os.name != "nt" else None
-        self._warn_if_fish(executable)
+
+        kwargs = {"cwd": self.wdir, "env": fix_env(None), "close_fds": True}
+
+        if os.name == "nt":
+            cmd = [self.cmd]
+            kwargs["shell"] = True
+        else:
+            # NOTE: when you specify `shell=True`, `Popen` [1] will default to
+            # `/bin/sh` on *nix and will add ["/bin/sh", "-c"] to your command.
+            # But we actually want to run the same shell that we are running
+            # from right now, which is usually determined by the `SHELL` env
+            # var. So instead, we compose our command on our own, making sure
+            # to include special flags to prevent shell from reading any
+            # configs and modifying env, which may change the behavior or the
+            # command we are running. See [2] for more info.
+            #
+            # [1] https://github.com/python/cpython/blob/3.7/Lib/subprocess.py
+            #                                                            #L1426
+            # [2] https://github.com/iterative/dvc/issues/2506
+            #                                           #issuecomment-535396799
+            kwargs["shell"] = False
+            executable = os.getenv("SHELL") or "/bin/sh"
+
+            self._warn_if_fish(executable)
+
+            opts = {"zsh": ["--no-rcs"], "bash": ["--no-profile", "--norc"]}
```

Comment:
no chances for the `fish` even this time? :)

Reply:
Nope https://github.com/fish-shell/fish-shell/issues/3288 :slightly_frowning_face: But it got me thinking that we could maybe fallback to `sh` or `bash` if we detect `fish`, but that would be pretty hacky and would break things if someone relies on fish-specific behaviour.  We could think about making that configurable, but I don't think it is worth it at this point. Might reconsider next time someone asks about fish.

Reply:
@efiop so there is no easy way to test any behavior to it is related to fish nor my config. it is sad.